src/modules/mc_pos_control/mc_pos_control_main.cpp

```diff
@@ -1517,23 +1588,45 @@ MulticopterPositionControl::task_main()
 				}
 
 			} else if (_control_mode.flag_control_manual_enabled
-				   && _vehicle_land_detected.landed) {
-				/* don't run controller when landed */
-				_reset_pos_sp = true;
-				_reset_alt_sp = true;
-				_do_reset_alt_pos_flag = true;
-				_mode_auto = false;
-				reset_int_z = true;
-				reset_int_xy = true;
+				   && (landed_for_sure || (_landed_locked && _acc_z_lp < 4.0f))) {
+				if (_manual.z >= _takeoff_throttle_thresh.get() || _takeoff_thrust_sp > FLT_EPSILON) {
+					if (fabsf(_takeoff_alt) < FLT_EPSILON) {
+						_takeoff_alt = _pos(2);
+					}
 
-				R.identity();
+					do_ground_takeoff(dt);
+					_att_sp.thrust = _takeoff_thrust_sp;
+
+				} else {
+					_att_sp.thrust = 0.0f;
+					reset_yaw_sp = true;
+				}
+
+				if (!_takeoff_jumped) {
+					/* don't run controller when landed */
+					_reset_pos_sp = true;
+					_reset_alt_sp = true;
+					_mode_auto = false;
+					reset_int_xy = true;
+					_alt_hold_engaged = false;
+					_pos_hold_engaged = false;
+
+				} else {
+					_pos_sp(2) = _takeoff_alt - _tk_alt_sp.get();
+					_takeoff_alt = 0.0f;
+					_reset_int_z = false;	// integrator has been set, do not reset!!!
+					_alt_hold_engaged = true;
+					_pos_hold_engaged = true;
+					_landed_locked = false;
+				}
 
 				_att_sp.roll_body = 0.0f;
 				_att_sp.pitch_body = 0.0f;
-				_att_sp.yaw_body = _yaw;
-				_att_sp.thrust = 0.0f;
-
+				R = matrix::Eulerf(_att_sp.roll_body, _att_sp.pitch_body, _att_sp.yaw_body);
```

Comment:
@tumbili: I followed your suggestion to use matrix lib. I actually have a PR for Julians refactor mc_pos_control, and then everything should work with matrix. At some point we should do the same with attitude controller